File: LeetCode/Hard/0127-word-ladder/0127-word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordList = set(wordList)

        if endWord not in wordList:
            return 0

        length = len(beginWord)
        ret = 1
        stack = [beginWord]
        idx = 0

        while idx < len(stack):
            ret += 1
            end = len(stack)
            
            for i in range(idx, end):
                idx += 1
                temp = stack[i]

                for j in range(length):
                    for v in string.ascii_lowercase:
                        val = temp[:j] + v + temp[j+1:]
                        if val in wordList:
                            if val == endWord:
                                return ret
                            stack.append(val)
                            wordList.remove(val)

        return 0

File: LeetCode/Hard/0127-word-ladder/0127-word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        HashSet<String> words = new HashSet(wordList);
        
        if (!words.contains(endWord)) return 0;
        
        ArrayList<String> stack = new ArrayList();
        stack.add(beginWord);
        int idx = 0;
        int ret = 1;
        
        while (idx < stack.size()){
            int end = stack.size();
            ++ret;
            
            for (int i = idx; i < end; ++i){
                ++idx;
                char[] temp = stack.get(i).toCharArray();
                
                for (int j = 0; j < temp.length; ++j){
                    char tempChar = temp[j];
                    for (int k = 0; k < 26; ++k){
                        temp[j] = (char) ('a' + k);
                        String val = new String(temp);

                        if (words.contains(val)){
                            if (val.equals(endWord)) return ret;
                            stack.add(val);
                            words.remove(val);
                        }
                    }
                    temp[j] = tempChar;
                }
            }
        }
        
        return 0;
    }
}